server/src/main/java/org/opensearch/search/backpressure/stats/SearchShardTaskStats.java

```diff
@@ -0,0 +1,128 @@
+
+    public SearchShardTaskStats(
+        long cancellationCount,
+        long limitReachedCount,
+        CancelledTaskStats lastCancelledTaskStats,
+        Map<TaskResourceUsageTrackerType, TaskResourceUsageTracker.Stats> resourceUsageTrackerStats
+    ) {
+        this.cancellationCount = cancellationCount;
+        this.limitReachedCount = limitReachedCount;
+        this.lastCancelledTaskStats = lastCancelledTaskStats;
+        this.resourceUsageTrackerStats = resourceUsageTrackerStats;
+    }
+
+    public SearchShardTaskStats(StreamInput in) throws IOException {
+        this(in.readVLong(), in.readVLong(), in.readOptionalWriteable(CancelledTaskStats::new), readResourceUsageTrackerStats(in));
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+
+        builder.startObject("resource_tracker_stats");
+        for (Map.Entry<TaskResourceUsageTrackerType, TaskResourceUsageTracker.Stats> entry : resourceUsageTrackerStats.entrySet()) {
+            builder.field(entry.getKey().getName(), entry.getValue());
+        }
+        builder.endObject();
+
+        builder.startObject("cancellation_stats")
+            .field("cancellation_count", cancellationCount)
+            .field("cancellation_limit_reached_count", limitReachedCount)
+            .field("last_cancelled_task", lastCancelledTaskStats)
+            .endObject();
+
+        return builder.endObject();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVLong(cancellationCount);
+        out.writeVLong(limitReachedCount);
+        out.writeOptionalWriteable(lastCancelledTaskStats);
+        out.writeMap(resourceUsageTrackerStats, (o, type) -> o.writeString(type.getName()), (o, stats) -> stats.writeTo(o));
+    }
+
+    private static Map<TaskResourceUsageTrackerType, TaskResourceUsageTracker.Stats> readResourceUsageTrackerStats(StreamInput in)
+        throws IOException {
+        int size = in.readVInt();
+        if (size == 0) {
+            return Collections.emptyMap();
+        }
+
+        MapBuilder<TaskResourceUsageTrackerType, TaskResourceUsageTracker.Stats> builder = new MapBuilder<>();
+
+        for (int i = 0; i < size; i++) {
+            TaskResourceUsageTrackerType type = TaskResourceUsageTrackerType.fromName(in.readString());
+            TaskResourceUsageTracker.Stats stats;
+
+            switch (type) {
+                case CPU_USAGE_TRACKER:
+                    stats = new CpuUsageTracker.Stats(in);
+                    break;
+                case HEAP_USAGE_TRACKER:
+                    stats = new HeapUsageTracker.Stats(in);
+                    break;
+                case ELAPSED_TIME_TRACKER:
+                    stats = new ElapsedTimeTracker.Stats(in);
+                    break;
+                default:
```

Comment:
Thanks for pointing out this scenario. We'll handle BWC just like we do in other places.

Here's an example when reading stats from an InputStream:
```java
switch (type) {
   case ELAPSED_TIME_TRACKER:
       stats = new ElapsedTimeTracker.Stats(in);
       break;
   case SOME_NEW_TRACKER:
       if (in.getVersion().onOrAfter(Version.V_2_5_0)) {
           stats = new SomeNewTracker.Stats(in);
       } else {
           stats = null;
       }
       break;
}
```

Writing to the OutputStream will require some modification. Instead of using the shorthand `out.writeMap(resourceUsageTrackerStats, (o, type) -> o.writeString(type.getName()), (o, stats) -> stats.writeTo(o));`, we'll have to replace it with the following:

```java
out.writeVInt(resourceUsageTrackerStats.size());
for (Map.Entry<TaskResourceUsageTrackerType, TaskResourceUsageTracker.Stats> entry : resourceUsageTrackerStats.entrySet()) {
    switch (entry.getKey()) {
        case ELAPSED_TIME_TRACKER:
            entry.getValue().writeTo(out);
            break;
        case SOME_NEW_TRACKER:
            if (out.getVersion().onOrAfter(Version.V_2_5_0)) {
                entry.getValue().writeTo(out);
            }
            break;
    }
}
```

Reply:
Thinking about this a bit more, I do see some issues with this still.

Suppose a StreamInput of a higher version is processed by a node of a lower version, it will be able to read the tracker type (the map's key of string type), but it cannot simply ignore the bytes immediately following it. So we must write the values in a deterministic order.

Can you confirm if my understanding is correct?

Reply:
I have made these changes now. Using `out.writeMap(...)` introduces non-determinism as keys could be unordered, I have now replaced it with manually writing the stats in a known order.

Any new trackers that get introduced in the future must write their output at the end of the StreamOutput along with the relevant version checks.